allel/stats/roh.py

```diff
@@ -0,0 +1,231 @@
+        contig_size = is_accessible.size
+
+    emission_mx = _derive_emission_mx(emission_px, p_accessible)
+
+    # initialize HMM
+    roh_hmm = hmm.MultinomialHMM(n_components=emission_px.size)
+
+    roh_hmm.n_symbols_ = 3
+    roh_hmm.startprob_ = start_prob
+    roh_hmm.transmat_ = transitions
+    roh_hmm.emissionprob_ = emission_mx
+
+    gsub = g.take([ix], axis=1)
+    hz = gsub.count_het(axis=1).astype("bool")
+
+    pred, obs = _predict_roh_state(roh_hmm, hz, pos, is_accessible, contig_size)
+    homozygous_windows = _get_state_windows(pred, state=0)
+
+    # filter by roh size
+    roh_sizes = np.diff(homozygous_windows, axis=1).flatten()
+    roh = np.compress(roh_sizes >= min_roh, homozygous_windows, axis=0)
+    froh = np.diff(roh, axis=1).sum() / contig_size
+
+    if plot:
+        import matplotlib.pyplot as plt
+        from allel.stats import windowed_statistic
+
+        window = np.arange(0, is_accessible.size, plotwinsize)
+        windows = np.vstack([window[:-1], window[1:]]).T
+
+        count_hz, win, count = windowed_statistic(pos=pos, values=hz,
+                                                  statistic=np.sum,
+                                                  windows=windows)
+
+        if is_accessible is not None:
+
+            # remove the last val as not a complete window
+            count_acc = np.add.reduceat(
+                is_accessible, window)[:-1]
+
+            # if accessible bases in a window < 1000, don't plot.
+            # b/c Leads to spiky data
+            count_acc[count_acc < 1000] = 0
+            count_hz[count_acc < 1000] = 0
+            heterozyg = count_hz/count_acc
+
+            assert not np.any(count_hz > count_acc), \
+                "heterozygosity exceeds 1.0. Check accessibility array input"
+        else:
+            heterozyg = count_hz/plotwinsize
+
+        x = win.mean(axis=1) * 1e-6
+
+        plt.figure(figsize=(6, 1.5))
+        plt.plot(x, heterozyg, "b-", label="Heterozygosity")
+        plt.ylabel('Heterozygosity')
+
+        for low, upp in roh * 1e-6:
+            plt.axvspan(low, upp, facecolor='g', alpha=0.5, edgecolor='none')
+
+        if is_accessible is not None:
+            inaccessible_windows = _get_state_windows(is_accessible)
+
+            for low, upp in (inaccessible_windows * 1e-6):
+                # don't plot inaccessible gaps < 5kb
+                if (upp - low) > .005:
+                    plt.axvspan(low, upp, facecolor='m', alpha=0.1,
+                                edgecolor='none')
+
+        plt.xlabel('Position (Mbp)')
+        plt.title(" ".join(['Called runs of homozygosity.',
+                            r'$F_{ROH}:$',
+                            "{0:.4f}".format(froh)]))
+        plt.tick_params(top="off", right="off")
+        plt.legend()
+
+    # return roh and froh
+    return roh, froh
+
+
+def _predict_roh_state(model, is_het, pos, is_accessible=None,
```

Comment:
Suggest to rename _mhmm_predict_roh_state to leave room for other roh-calling functions in future.